modules/management-console-agent/src/main/java/org/apache/ignite/agent/Agent.java

```diff
@@ -0,0 +1,451 @@
+
+            mgr.stop(true);
+        }
+        catch (InterruptedException e) {
+            if (log.isDebugEnabled())
+                log.debug("Caught interrupted exception: " + e);
+
+            mgr.stop(true);
+
+            Thread.currentThread().interrupt();
+        }
+        catch (TimeoutException ignored) {
+            connect0();
+        }
+        catch (ExecutionException e) {
+            mgr.stop(true);
+
+            if (X.hasCause(e, ConnectException.class, UpgradeException.class, EofException.class, ConnectionLostException.class)) {
+                if (disconnected.compareAndSet(false, true))
+                    log.error("Failed to establish websocket connection with Management Console: " + curSrvUri);
+
+                connect0();
+            }
+            else
+                log.error("Failed to establish websocket connection with Management Console: " + curSrvUri, e);
+        }
+        catch (Exception e) {
+            log.error("Failed to establish websocket connection with Management Console: " + curSrvUri, e);
+        }
+    }
+
+    /**
+     * Connect to backend.
+     */
+    private void connect() {
+        if (!cfg.isEnabled()) {
+            log.info("Skip start Management Console agent on coordinator, because it was disabled in configuration");
+
+            return;
+        }
+
+        log.info("Starting Management Console agent on coordinator");
+
+        mgr = new WebSocketManager(ctx);
+        sessionRegistry = new SessionRegistry(ctx);
+        clusterSrvc = new ClusterService(ctx, mgr);
+        tracingSrvc = new TracingService(ctx, mgr);
+        metricSrvc = new MetricsService(ctx, mgr);
+        evtSrvc = new EventsService(ctx, mgr);
+        nodeConfigurationSrvc = new NodeConfigurationService(ctx, mgr);
+        actSrvc = new ActionService(ctx, mgr);
+        cacheSrvc = new CacheService(ctx, mgr);
+
+        evtsExporter.addGlobalEventListener();
+
+        connectPool =
+            (ThreadPoolExecutor) Executors.newFixedThreadPool(1, new CustomizableThreadFactory("mgmt-console-connection-"));
+
+        connectPool.submit(this::connect0);
+
+        ctx.event().enableEvents(NOT_ENABLED_EVTS);
+    }
+
+    /**
+     * @return Agent configuration.
+     */
+    private ManagementConfiguration readFromMetaStorage() {
+        if (metaStorage == null)
+            return new ManagementConfiguration();
+
+        ctx.cache().context().database().checkpointReadLock();
+
+        ManagementConfiguration cfg = null;
+
+        try {
+            cfg = metaStorage.read(MANAGEMENT_CFG_META_STORAGE_PREFIX);
+        }
+        catch (IgniteCheckedException e) {
+            log.warning("Failed to read management configuration from meta storage!");
+        }
+        finally {
+            ctx.cache().context().database().checkpointReadUnlock();
+        }
+
+        return cfg != null ? cfg : new ManagementConfiguration();
```

Comment:
`cfg` can be `null` if exception was thrown. It seems that it is error because, for example, we can have broken data/files. But agent  will correctly started with new configuration.

Reply:
From the user's point of view, it will be better to start the agent with the default configuration and permit change it by control script than the management utility will be blocked.

Reply:
@nva If metastore can't read data (throws exception) than data is broken and this exception should be passed on upper level. There is no need in starting agent if node can't work with data.

Reply:
Throw unchecked exception with throw U.convertException(e);